### PR TITLE
deprecate Bazel builder in @angular/bazel

### LINKED ISSUE #19058: Tracking issue: Angular BuildTools Convergence (ABC)
We are open-sourcing the toolchain we use internally at Google to build Angular apps at scale.

One-pager: http://g.co/ng/abc

This issue will be used to give updates on progress.

FAQ: how does this relate to Angular CLI? We will evaluate this toolchain with some early adopters first, if it goes well, we will evaluate whether to use it inside the CLI. So long as you haven't ejected from the CLI, such a change should be transparent to users.

### PR DESCRIPTION
**Pull request by kyliau**
This PR removes the schematics for Bazel builder from `@angular/bazel`.
It consists of a few commits:

1. Add migration doc for existing users and remove Bazel guide from angular.io.
2. Remove all `@angular/bazel` schematics code, mainly `ng-add` and `ng-new`, and the Bazel builder.
3. Remove "Building with Bazel" section from angular.io
4. Remove `aio/content/guide/bazel.md` from pullapprove
5. Remove `bazel-schematics` integration test
6. Redirect `guide/bazel` to deprecation doc in Angular repo

PR closes https://github.com/angular/angular/issues/19058

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: packages/bazel/src/schematics/README.md [hunk 1]
```diff
@@ -1,136 +1,46 @@
...
+The schematics for Bazel builder have been in Angular Labs for some time,
+but over the course of its evaluation the Angular team has found that they did
+not meet our expectation of providing a smooth experience to get users onboard.
 
-## Requirements
+There are multiple factors:
 
-To create a new Angular project that builds with Bazel + CLI, the following
-packages have to be installed.
+1. Bazel ecosystem for Javascript is still evolving at a rapid pace.
```
Comment (IgorMinar) [inline packages/bazel/src/schematics/README.md:11]:
```suggestion
1. Bazel ecosystem for the Web is still evolving at a rapid pace.
```

File: packages/bazel/src/schematics/README.md [hunk 2]
```diff
@@ -1,136 +1,46 @@
...
+2. The introduction of Ivy compiler enables new ways to use Bazel in a faster
```
Comment (IgorMinar) [inline packages/bazel/src/schematics/README.md:12]:
```suggestion
2. The introduction of the Angular Ivy compiler enables new ways to use Bazel in a faster
```

File: packages/bazel/src/schematics/README.md [hunk 3]
```diff
@@ -1,136 +1,46 @@
...
+   and more efficient manner.
+3. Feature parity with Angular CLI is difficult to achieve.
```
Comment (IgorMinar) [inline packages/bazel/src/schematics/README.md:14]:
```suggestion
3. Feature parity with Webpack based Angular CLI is currently difficult to achieve without trade offs that would not be acceptable for many Angular users.
```

File: packages/bazel/src/schematics/README.md [hunk 4]
```diff
@@ -1,136 +1,46 @@
...
-## WARNING
+Bazel builder for Angular CLI has been deprecated.
 
-Schematics in `@angular/bazel` is still highly experimental as of January 2019,
-please use with caution. For feedbacks and comments, please open an issue on
-GitHub and ping [@mgechev](https://github.com/mgechev) or
-[@kyliau](https://github.com/kyliau).
+The schematics for Bazel builder have been in Angular Labs for some time,
+but over the course of its evaluation the Angular team has found that they did
+not meet our expectation of providing a smooth experience to get users onboard.
```
Comment (IgorMinar) [inline packages/bazel/src/schematics/README.md:7]:
```suggestion
The schematics for Bazel builder have been in Angular Labs for some time. Via labs we have been able to experiment with wrapping Bazel into the CLI and make it's use in the CLI invisible to developers using Angular CLI.

Through this experiment we, the Angular team have found that wrapping Bazel into Angular CLI does currently not meet our expectation of providing a great experience to developers using Angular.

```

File: packages/bazel/src/schematics/README.md [hunk 5]
```diff
@@ -1,136 +1,46 @@
 # Schematics for Bazel
 
-## WARNING
+Bazel builder for Angular CLI has been deprecated.
```
Comment (IgorMinar) [inline packages/bazel/src/schematics/README.md:3]:
```suggestion
Bazel builder for Angular CLI has been deprecated.

**tl;dr: We have deprecated the the Bazel builder and schematics for Angular CLI. As of Angular v10 we removed the ability to call `ng add @angular/bazel` to convert existing Angular CLI projects to use Bazel. We believe that some Angular projects can greatly benefit from using Bazel - these projects should use Bazel Web directly as documented in the [canonical example in the Bazel Web repo](https://github.com/bazelbuild/rules_nodejs/tree/master/examples/angular).**
```

File: aio/content/guide/bazel.md [hunk 6]
```diff
@@ -1,122 +0,0 @@
-# Building with Bazel
```
Comment (IgorMinar) [inline aio/content/guide/bazel.md:1]: we can't just remove this link. We'll need to somehow point people to the readme.md with the deprecation notice. We can do this by setting up a redirect via https://github.com/angular/angular/blob/master/aio/firebase.json

**Review by IgorMinar — CHANGES_REQUESTED** [review]
I think that the README.md needs some more work and likely devrel input. Can you please get @mgechev and @StephenFluin to weigh in on the wording? We'll also need @aikidave to look at this before it hits angular.io in June.

In general we don't want to freak people out with this removal. It's not that we don't like Bazel (some of us very much do! 😃), we just found out that our current approach to jam bazel into Angular CLI is not the right one, and we need to pivot. Instead of wrapping Bazel into Angular CLI we found a way how bazel and Angular CLI can coexist for some workflows and for others it's actually better if Bazel wraps Angular CLI.

In order to enable this pivot, we must deprecate the current integration. For people that depend on Angular CLI + Bazel, they can keep on building their apps with the existing rules (`ng_module`, `ng_package`, etc), but we don't want new projects to onboard in this way any more.

In the future, we'll most likely remove all of these custom rules and replace them with direct integration from bazel to `ngc`, etc. But we are nowhere close to that. We'll revisit how this stuff should work once we remove View Engine from our code base and Ivy is the only compilation pipeline out there.

File: aio/content/guide/bazel.md (hunk 6 above)
Comment (kyliau) [reply aio/content/guide/bazel.md:1]: Redirect added for `guide/bazel`, will point to https://github.com/angular/angular/blob/master/packages/bazel/src/schematics/README.md

File: packages/bazel/src/schematics/README.md [hunk 7]
```diff
@@ -1,136 +1,49 @@
...
 
-Schematics in `@angular/bazel` is still highly experimental as of January 2019,
-please use with caution. For feedbacks and comments, please open an issue on
-GitHub and ping [@mgechev](https://github.com/mgechev) or
-[@kyliau](https://github.com/kyliau).
+**tl;dr: We have deprecated the the Bazel builder and schematics for Angular CLI. As of Angular v10 we removed the ability to call `ng add @angular/bazel` to convert existing Angular CLI projects to use Bazel. We believe that some Angular projects can greatly benefit from using Bazel - these projects should use Bazel Web directly as documented in the [canonical example in the Bazel Web repo](https://github.com/bazelbuild/rules_nodejs/tree/master/examples/angular).**
```
Comment (mgechev) [inline packages/bazel/src/schematics/README.md:5]:
```suggestion
We believe that a lot of projects would benefit from directly using Bazel, the same way we do in Angular. Validating the experimental Bazel schematics for the Angular CLI for the past year as part of Angular Labs, we concluded that they don't implement the most optimal approach. In v10 we're deprecating `@angular/bazel` and encouraging developers to explore using Bazel directly following the [canonical example](https://github.com/bazelbuild/rules_nodejs/tree/master/examples/angular).
```

File: packages/bazel/src/schematics/README.md [hunk 8]
```diff
@@ -1,136 +1,49 @@
...
 
-## Requirements
+The schematics for Bazel builder have been in Angular Labs for some time. Via labs we have been able to experiment with wrapping Bazel into the CLI and make it's use in the CLI invisible to developers using Angular CLI.
```
Comment (mgechev) [inline packages/bazel/src/schematics/README.md:7]:
```suggestion
Using the Bazel builder we abstracted the build file management and tooling orchestration, making Bazel invisible for developers.
```

File: packages/bazel/src/schematics/README.md [hunk 9]
```diff
@@ -1,136 +1,49 @@
...
 
-To create a new Angular project that builds with Bazel + CLI, the following
-packages have to be installed.
+Through this experiment we, the Angular team have found that wrapping Bazel into Angular CLI does currently not meet our expectation of providing a great experience to developers using Angular.
```
Comment (mgechev) [inline packages/bazel/src/schematics/README.md:9]:
```suggestion
Through this experiment we found that wrapping Bazel into Angular CLI does currently not meet our expectation of providing a great experience to developers.
```

File: packages/bazel/src/schematics/README.md [hunk 10]
```diff
@@ -1,136 +1,49 @@
...
 
-|     Package    | Minimum Version
-|----------------|-----------------
-| @angular/cli   | v8.0.x
-| @angular/bazel | v8.0.x
 
-The `@angular/bazel` package contains schematics to generate necessary Bazel
-build files.
+There are multiple factors:
```
Comment (mgechev) [inline packages/bazel/src/schematics/README.md:12]:
```suggestion
There are a few reasons we're taking this path:
```

File: packages/bazel/src/schematics/README.md [hunk 11]
```diff
@@ -1,136 +1,49 @@
...
 
-If the packages are not on your system yet, install them with the following
-commands:
+1. Bazel ecosystem for the Web is still evolving at a rapid pace.
+2. The introduction of the Angular Ivy compiler enables new ways to use Bazel in a faster
+   and more efficient manner.
+3. Feature parity with Webpack based Angular CLI is currently difficult to achieve without trade offs that would not be acceptable for many Angular users.
```
Comment (mgechev) [inline packages/bazel/src/schematics/README.md:17]:
```suggestion
3. Feature parity with Webpack based Angular CLI is currently difficult to achieve without trade-offs that would not be acceptable for many Angular users.
```

File: packages/bazel/src/schematics/README.md [hunk 12]
```diff
@@ -1,136 +1,49 @@
...
-| @angular/cli   | v8.0.x
-| @angular/bazel | v8.0.x
 
-The `@angular/bazel` package contains schematics to generate necessary Bazel
-build files.
+There are multiple factors:
 
-If the packages are not on your system yet, install them with the following
-commands:
+1. Bazel ecosystem for the Web is still evolving at a rapid pace.
```
Comment (mgechev) [inline packages/bazel/src/schematics/README.md:14]: I'd also add that hiding the build graph defeats the purpose of using Bazel.

File: packages/bazel/src/schematics/README.md [hunk 13]
```diff
@@ -1,136 +1,49 @@
...
-## Create a Bazel-managed Angular project
-
-Create a new project using `@angular/bazel` schematics for `ng new`.
+For users who are currently using Bazel builder, there are a few migration
+options.
 
-```
-$ ng new demo --collection=@angular/bazel --defaults
-```
+## View Engine
```
Comment (mgechev) [inline packages/bazel/src/schematics/README.md:22]:
```suggestion
```

File: packages/bazel/src/schematics/README.md [hunk 14]
```diff
@@ -1,136 +1,49 @@
...
 
-In addition to the regular files generated by CLI, the following files that are
-specific to a Bazel workspace are also created.
+If your application is using View Engine, you could leave the Bazel files
```
Comment (mgechev) [inline packages/bazel/src/schematics/README.md:24]:
```suggestion
```

File: packages/bazel/src/schematics/README.md [hunk 15]
```diff
@@ -1,136 +1,49 @@
...
+in your workspace, and manage them manually.
```
Comment (mgechev) [inline packages/bazel/src/schematics/README.md:25]:
```suggestion
```

File: packages/bazel/src/schematics/README.md [hunk 16]
```diff
@@ -1,136 +1,49 @@
...
 
 ```
```
Comment (mgechev) [inline packages/bazel/src/schematics/README.md:27]:
```suggestion
```

File: packages/bazel/src/schematics/README.md [hunk 17]
```diff
@@ -1,136 +1,49 @@
...
-CREATE demo/WORKSPACE (2951 bytes)
-CREATE demo/.bazelignore (18 bytes)
-CREATE demo/.bazelrc (828 bytes)
-CREATE demo/e2e/BUILD.bazel (1230 bytes)
-CREATE demo/e2e/protractor.on-prepare.js (1101 bytes)
-CREATE demo/src/BUILD.bazel (2626 bytes)
-CREATE demo/src/initialize_testbed.ts (432 bytes)
-CREATE demo/src/main.dev.ts (185 bytes)
-CREATE demo/src/main.prod.ts (249 bytes)
+ng build --leaveBazelFilesOnDisk
```
Comment (mgechev) [inline packages/bazel/src/schematics/README.md:28]:
```suggestion
```

File: packages/bazel/src/schematics/README.md [hunk 18]
```diff
@@ -1,136 +1,49 @@
...
 ```
```
Comment (mgechev) [inline packages/bazel/src/schematics/README.md:29]:
```suggestion
```

File: packages/bazel/src/schematics/README.md [hunk 19]
```diff
@@ -1,136 +1,49 @@
...
 
-Note that in a Bazel-managed project, there is a Bazel WORKSPACE file and a few BUILD.bazel files.
-There are also some files specific to a Bazel-managed project, namely `main.dev.ts` and `main.prod.ts`.
-This is because all Angular projects built with Bazel must be AOT only.
-In a Bazel project, `main.ts` generated by CLI is not used.
-
-By default, `ng new` for Bazel does not perform `yarn install`.
-This is because the `node_modules` are managed by Bazel and it is Bazel's
-responsibility to perform the install.
+## Ivy
```
Comment (mgechev) [inline packages/bazel/src/schematics/README.md:31]:
```suggestion
## Migrate from `@angular/bazel`
```

File: packages/bazel/src/schematics/README.md [hunk 20]
```diff
@@ -1,136 +1,49 @@
...
 
-Next, let's try to build the project using Bazel.
-All existing `ng` commands would work as before.
+If your application is using Ivy, you should also leave the Bazel files on disk.
```
Comment (mgechev) [inline packages/bazel/src/schematics/README.md:33]:
```suggestion
You could leave the Bazel files in your workspace, and manage them manually:
```

File: packages/bazel/src/schematics/README.md [hunk 21]
```diff
@@ -1,136 +1,49 @@
...
-# equivalent to running `ng build`.
-yarn
-ng build
-ng serve
+ng build --leaveBazelFilesOnDisk
 ```
 
-If you encounter a warning about version mismatch, update `ANGULAR_VERSION` in
-the WORKSPACE file to match the version of `@angular/bazel` installed from NPM.
+Once that is done, switch to using the latest [`ng_ts_library`](https://github.com/bazelbuild/rules_nodejs/blob/master/examples/angular/tools/angular_ts_library.bzl) rule.
```
Comment (mgechev) [inline packages/bazel/src/schematics/README.md:39]:
```suggestion
If you're using Ivy, in your build files replace `ng_module` with the [`ng_ts_library`](https://github.com/bazelbuild/rules_nodejs/blob/master/examples/angular/tools/angular_ts_library.bzl) rule. This new rule leverages ngtsc plugin supported by `ts_library`, and it is much faster.
```

File: packages/bazel/src/schematics/README.md [hunk 22]
```diff
@@ -1,136 +1,49 @@
...
+This new rule leverages ngtsc plugin supported by `ts_library`, and it is much
```
Comment (mgechev) [inline packages/bazel/src/schematics/README.md:40]:
```suggestion
```

File: packages/bazel/src/schematics/README.md [hunk 23]
```diff
@@ -1,136 +1,49 @@
...
+faster.
```
Comment (mgechev) [inline packages/bazel/src/schematics/README.md:41]:
```suggestion
```

**Review by alan-agius4 — CHANGES_REQUESTED** [review]
Can you please remove the following entries from package.json
https://github.com/angular/angular/blob/3a4e1b8f860eb7449144a0866de750d9e18d2501/packages/bazel/package.json#L54-L57

File: packages/bazel/src/schematics/README.md (hunk 11 above)
Comment (alan-agius4) [reply packages/bazel/src/schematics/README.md:17]: Maybe mention that it's still possible to use Webpack and Angular CLI with Bazel. Though you'd don't have all the benefits of using Bazel. It's more of an easier version of the setup.

Example using Angular CLI + Bazel with Architect
https://github.com/bazelbuild/rules_nodejs/tree/master/examples/angular_bazel_architect

File: aio/firebase.json [hunk 24]
```diff
@@ -18,6 +18,7 @@
       {"type": 301, "source": "/docs/*/latest/quickstart.html", "destination": "/start"},
       {"type": 301, "source": "/docs/*/latest/guide/server-communication.html", "destination": "/guide/http"},
       {"type": 301, "source": "/docs/*/latest/guide/style-guide.html", "destination": "/guide/styleguide"},
+      {"type": 301, "source": "/guide/bazel", "destination": "https://github.com/angular/angular/blob/master/packages/bazel/src/schematics/README.md"},
```
Comment (gkalpak) [inline aio/firebase.json:21]: In order for redirects to work correctly, you also need to add corresponding entries in `ngsw-config.json`:

https://github.com/angular/angular/blob/73499806d1b679e81b3c1a0e763d02cd525b6b78/aio/ngsw-config.json#L101-L103

**Comment by kyliau** [comment]
@alan-agius4 and @gregmagolan, in the deprecation doc we mentioned that orchestrating CLI with Bazel is better than wrapping Bazel within CLI. Do you think it’s worth to mention the current prototype we have?
https://github.com/bazelbuild/rules_nodejs/tree/master/examples/angular_bazel_architect

**Comment by kyliau** [comment]
All comments addressed, please take a look again, thank you!

**Comment by alan-agius4** [comment]
> @alan-agius4 and @gregmagolan, in the deprecation doc we mentioned that orchestrating CLI with Bazel is better than wrapping Bazel within CLI. Do you think it’s worth to mention the current prototype we have?
https://github.com/bazelbuild/rules_nodejs/tree/master/examples/angular_bazel_architect

I would, but I'd like to hear more about this from @gregmagolan.

**Comment by alexeagle** [comment]
@gregmagolan is pretty busy lately, maybe I can step in and help? I think pointing to that example is great.

I have some content I've worked up in the past to go along with this migration. If you loop me into the discussion I'd like to help contribute that.

At what point should there be a builder in a `@bazel/angular` package? @mgechev you probably have a better idea than anyone about how many people rely on `ng` as a frontend to the bazel command.

File: packages/bazel/src/schematics/README.md [hunk 25]
```diff
@@ -1,136 +1,50 @@
...
 
-## WARNING
+Bazel builder for Angular CLI has been deprecated in version 10.
 
-Schematics in `@angular/bazel` is still highly experimental as of January 2019,
-please use with caution. For feedbacks and comments, please open an issue on
-GitHub and ping [@mgechev](https://github.com/mgechev) or
-[@kyliau](https://github.com/kyliau).
+We believe that a lot of projects would benefit from using Bazel, the same way
+we do in Angular repository. We still believe that is true in many cases.
```
Comment (aikidave) [inline packages/bazel/src/schematics/README.md:6]:
```suggestion
we do in the Angular repository. We still believe that is true in many cases.
```

**Review by aikithoughts — APPROVED** [review]
One minor nit. Otherwise, LGTM!

**Review by alan-agius4 — APPROVED** [review]

**Review by alan-agius4 — APPROVED** [review]
LGTM, maybe we should indeed add a link to https://github.com/bazelbuild/rules_nodejs/tree/master/examples/angular_bazel_architect as @alexeagle mentioned.

**Comment by alexeagle** [comment]
I'll update the README in https://github.com/bazelbuild/rules_nodejs/tree/master/examples/angular so if you point users there, it will still help disambiguate.

**Comment by alexeagle** [comment]
What is going to happen with `bazel.angular.io`?

Since my angular.io account was taken away, I can't tell who has registered for support on the form https://goo.gl/forms/Tv8lMdH0N3D4yo8Q2. Has anyone replied to them? Will you close out the form so there are no new submissions?

You should probably suggest that ppl ask questions in slack.bazel.build in the `#angular` channel.

I assume it is time to close the tracking issue https://github.com/angular/angular/issues/19058?

I have a blog post I'd like to publish, I sent over Slack to avoid public disclosure before it's reviewed.

Finally, I think we should have at least a Builder in `@bazel/angular` as part of this migration, so it's still possible to run the `ng` command, WDYT?

**Review by IgorMinar — APPROVED** [review]
lgtm now. thanks for all the updates! 

Can you please add a note about this removal to https://next.angular.io/guide/deprecations#removed as well? (and link to the readme.md you wrote) thanks!

Reviewed-for: global-approvers

**Comment by kyliau** [comment]
> lgtm now. thanks for all the updates!
> 
> Can you please add a note about this removal to https://next.angular.io/guide/deprecations#removed as well? (and link to the readme.md you wrote) thanks!
> 
> Reviewed-for: global-approvers

done!

**Comment by IgorMinar** [comment]
@alexeagle thanks for chiming in. 

> What is going to happen with bazel.angular.io?

we'll redirect it to the readme.md updated in this PR.

> Since my angular.io account was taken away, I can't tell who has registered for support on the form https://goo.gl/forms/Tv8lMdH0N3D4yo8Q2. Has anyone replied to them? Will you close out the form so there are no new submissions?

@kyliau and @mgechev will follow up on this.

> You should probably suggest that ppl ask questions in slack.bazel.build in the #angular channel.
can you send a follow up PR to update the readme.md once this PR lands. Can you also update the Angular example readme we are linking to with that info?

> I assume it is time to close the tracking issue #19058?
yes

> I have a blog post I'd like to publish, I sent over Slack to avoid public disclosure before it's reviewed.
I'm happy to review but I can't find it. Can you email it to me plz?

> Finally, I think we should have at least a Builder in @bazel/angular as part of this migration, so it's still possible to run the ng command, WDYT?
Does that still require additional changes to the Angular CLI or have we landed all the changes that unblock `@bazel/angular`?

Will you own and maintain `@bazel/angular` or who will?

Thanks again for helping out with this cleanup.

**Comment by alexeagle** [comment]
okay, we might want to deploy the example app somewhere else, I'll check w/ #bazel-web-team on bazel slack.

Minko reviewed my post already, let me know directly if you also want the draft link.

I ported the builder to @bazel/angular package already https://github.com/bazelbuild/rules_nodejs/tree/master/packages/angular
and will work w/ @gregmagolan to get a release of that (probably a 1.6.x of rules_nodejs) and yeah we'll own it along with the rest of rules_nodejs maintainership. Note it's just the builder, no schematics code, so there is very little interaction with CLI. (no `ng add` support)

**Comment by mgechev** [comment]
>Since my angular.io account was taken away, I can't tell who has registered for support on the form https://goo.gl/forms/Tv8lMdH0N3D4yo8Q2. Has anyone replied to them? Will you close out the form so there are no new submissions?

We have only one entry for 2020.
